### DISCUSSION TITLE
Compatibility of Plugins with Local LLM (Ollama)

### DISCUSSION THREAD

**Discussion by forge-oscar-caballero** [lobehub/lobehub#1538] [General Question] [answered]
Hi, there's a little hiccup with LobeChat when using plugins with Local LLMs (Ollama), plugins seems to work only with OpenAI models.

<img width="1015" alt="Screenshot 2024-03-11 at 9 53 46 AM" src="https://github.com/lobehub/lobe-chat/assets/100241997/fcb14a51-766c-41f6-a51e-87954ea95507">

Right now, if you're using a local model with Ollama in LobeChat, you can't use plugins, the extensions icon will be disabled. They only work with OpenAI models.

Is there a way to enable this?

**Comment by Terramoto** [comment]
Hey,

From what i read Ollama API does not follow the chatgpt api structure so you would need to setup Ollama in the OpenAI configuration panel using the http://127.0.0.1:11434/v1. Then create the models you have available in Ollama and enable the function calling:
![image](https://github.com/lobehub/lobe-chat/assets/5443354/422b4125-f4a9-4c3d-a2cd-c9d46b914a5f)
![image](https://github.com/lobehub/lobe-chat/assets/5443354/e5708872-988d-42b6-ba09-80f3b6f8b78d)

Even though some models seem to be able to detect the plugins available (or maybe the plugins are coming from the trained data), it doesn't seem to be able to call the functions. Does anyone know what the structure should be?

If i ask to crawl a website or get the weather, the data is outdated or is generated by the model.

I can get it to send the request without being formatted, but i'm guessing im missing how the returned schema is converted to an actual API call for the plugin?
![image](https://github.com/lobehub/lobe-chat/assets/5443354/4899fcd5-ff15-4a8c-b940-8f0cf4eee0f3)

Or is ollama expected to POST the API call to an endpoint?

EDIT: Looking at one of the plugins code on lobe-chat, it looks like these are intended to be called by ChatGPT API in which case it will not work with Ollama. It would need local scripts and something to interpret function call "tags" and associate the call with the script to action it.

**Comment by arvinxx** [comment]
ollama currently don't support function calling.

Refs: https://github.com/lobehub/lobe-chat/issues/1364#issuecomment-2081416085

**Reply by gianlazz** [reply]
This doesn't look to be the case any longer as of July. This should probably be reassessed.
https://ollama.com/blog/tool-support

**Reply by arvinxx** [reply, marked as answer]
Actually we have supported streaming tools calling a few months ago at [1.35.0](https://github.com/lobehub/lobe-chat/releases/tag/v1.35.0). You can try with models like qwen2.5
